plugins/ssr/ssrfwd.ml

```diff
@@ -340,13 +340,38 @@ let intro_lock ipats =
   let hnf' = Proofview.numgoals >>= fun ng ->
              Proofview.tclDISPATCH
                (ncons (ng - 1) ssrsmovetac @ [Proofview.tclUNIT ()]) in
+  let nomatch_vars env sigma eredex =
+    let default () =
+      ppdebug(lazy Pp.(str"under: warning (not an evar-redex):" ++
+                         pr_econstr_env env sigma eredex));
+      sigma, eredex in
+    let sigma, nomatch =
+      Ssrcommon.mkSsrmatchingConst "nomatch" env sigma in
+    if EConstr.isApp sigma eredex then
+      let () = ppdebug(lazy Pp.(str"under: an evar-redex: " ++
+                         pr_econstr_env env sigma eredex)) in
+      let (e, args) = EConstr.destApp sigma eredex in
+      if is_app_evar sigma e && Array.length args >= 1 then
+        let protect sigma t =
+          let sigma, ty = Typing.type_of env sigma t in
+          sigma, EConstr.mkApp(nomatch, [|ty; t|]) in
+        let sigma, largs = Array.fold_left
+                             (fun (sigma, list) arg ->
+                               let (sigma2, arg') = protect sigma arg in
+                               sigma2, arg' :: list)
+                             (sigma, []) args in
+        let args = Array.of_list @@ List.rev @@ largs in
+        sigma, EConstr.mkApp(e, args)
```

Comment:
I'm not so sure I get what you are doing here.
In particular it seems you generate `?e (nomatch x1) .. (nomatch xn)` while you should generate `(nomatch (?e x1 .. xn))`, that is you should wrap the entire evar, not all its arguments. What you do correctly prevents `xi` from being matched (hence the surjective_pairing example works) but seems more complicated than needed.

Reply:
You should also use `Retyping` to typecheck the (applied) evar.

Reply:
Hi @gares,

thanks for your feedback!

> In particular it seems you generate `?e (nomatch x1) .. (nomatch xn)`

exactly; in fact in the first version of the code (before commit 2ebaa79886504a92a0b00c9d6ab85742c876a8a9) I was generating a `(nomatch (?e x1 .. xn))` term with a much shorter code indeed, but in this case the under tactic was not working anymore, as the enclosing `Tactics.New.refine ~typecheck:true (fun sigma ->` seemed to eat the evar; probably due to the specific shape of `(nomatch (?e x1 .. xn))`…

So even if the current code generating `?e (nomatch x1) .. (nomatch xn)` is a bit longer, are you OK with this strategy (tagging each argument of the evar) if it's feasible to make it work?

Reply:
> seemed to eat the evar

What do you mean exactly?

Reply:
just that the evar that corresponds to the "second fonction in the context lemma" was unexpectedly instantiated (to the identity here):

namely if I revert the commit 2ebaa79886504a92a0b00c9d6ab85742c876a8a9 from this PR, recompile the plugin and relaunch the first test of `test-suite/ssr/under.v`:

```coq
Lemma test_big_nested_1 (F G : nat -> nat) (m n : nat) :
  \sum_(0 <= i < m) \sum_(0 <= j < n | odd (j * 1)) (i + j) =
  \sum_(0 <= i < m) \sum_(0 <= j < n | odd j) (j + i).
Proof.
Set Printing All.
under eq_bigr => i Hi.
Set Printing All.
```

then I get the subgoal:
```
  F, G : forall _ : nat, nat
  m, n, i : nat
  Hi : is_true true
  ============================
  Under_rel nat (@eq nat)
    (@BigOp.bigop nat nat O (index_iota O n)
       (fun j : nat => @BigBody nat nat j addn (odd (muln j (S O))) (addn i j)))
    (ssrmatching.nomatch nat i)
```

instead of something like
```
  F, G : forall _ : nat, nat
  m, n, i : nat
  Hi : is_true true
  ============================
  Under_rel nat (@eq nat)
    (@BigOp.bigop nat nat O (index_iota O n)
       (fun j : nat => @BigBody nat nat j addn (odd (muln j (S O))) (addn i j)))
    (ssrmatching.nomatch nat (?Goal i))
```

but what is a bit strange, is that this issue does not occur in this other example below (adapted from the minimal-working-example from #11118):

```coq
From Coq Require Import ssreflect.
From Coq Require Import List.
Set Debug SsrMatching.
Lemma test1 (A: list (nat * nat)) (f: nat -> nat) :
  map (fun (a: nat * nat) => fst (let '(x, y) := a in (f x, f y))) A =
  map (fun (a: nat * nat) => f (fst a)) A.
Proof.
under map_ext => a.
Set Printing All.
```

which yields:
```
  A : list (prod nat nat)
  f : forall _ : nat, nat
  a : prod nat nat
  ============================
  Under_rel nat (@eq nat)
    (@fst nat nat match a return (prod nat nat) with
                  | pair x y => @pair nat nat (f x) (f y)
                  end) (ssrmatching.nomatch nat (?Goal a))
```
(here, the ?Goal evar is kept uninstantiated…)

Reply:
just that the evar that corresponds to the "second fonction in the context lemma" was unexpectedly instantiated (to the identity here):

namely if I revert the commit 2ebaa79886504a92a0b00c9d6ab85742c876a8a9 from this PR, recompile the plugin and relaunch the first test of `test-suite/ssr/under.v`:

```coq
Lemma test_big_nested_1 (F G : nat -> nat) (m n : nat) :
  \sum_(0 <= i < m) \sum_(0 <= j < n | odd (j * 1)) (i + j) =
  \sum_(0 <= i < m) \sum_(0 <= j < n | odd j) (j + i).
Proof.
under eq_bigr => i Hi.
Set Printing All.
```

then I get the subgoal:
```
  F, G : forall _ : nat, nat
  m, n, i : nat
  Hi : is_true true
  ============================
  Under_rel nat (@eq nat)
    (@BigOp.bigop nat nat O (index_iota O n)
       (fun j : nat => @BigBody nat nat j addn (odd (muln j (S O))) (addn i j)))
    (ssrmatching.nomatch nat i)
```

instead of something like
```
  F, G : forall _ : nat, nat
  m, n, i : nat
  Hi : is_true true
  ============================
  Under_rel nat (@eq nat)
    (@BigOp.bigop nat nat O (index_iota O n)
       (fun j : nat => @BigBody nat nat j addn (odd (muln j (S O))) (addn i j)))
    (ssrmatching.nomatch nat (?Goal i))
```

but what is a bit strange, is that this issue does not occur in this other example below (adapted from the minimal-working-example from #11118):

```coq
From Coq Require Import ssreflect.
From Coq Require Import List.
Set Debug SsrMatching.
Lemma test1 (A: list (nat * nat)) (f: nat -> nat) :
  map (fun (a: nat * nat) => fst (let '(x, y) := a in (f x, f y))) A =
  map (fun (a: nat * nat) => f (fst a)) A.
Proof.
under map_ext => a.
Set Printing All.
```

which yields:
```
  A : list (prod nat nat)
  f : forall _ : nat, nat
  a : prod nat nat
  ============================
  Under_rel nat (@eq nat)
    (@fst nat nat match a return (prod nat nat) with
                  | pair x y => @pair nat nat (f x) (f y)
                  end) (ssrmatching.nomatch nat (?Goal a))
```
(here, the ?Goal evar is kept uninstantiated…)